Automatically Get Arguments for Endpoints & Compatibility with Flask-Classy

nickw444 opened the pull request:
These commits bring the following functionality:

## Automatically get arguments for endpoints which take arguments by default
Allows users to use @menu.register_menu on endpoints with routes such as /<int:id>/endpoint1,  /<int:id>/endpoint2, /<int:id>/endpoint3, which need to link between each other (say to provide a user profile, with links between sections on a specific user profile.).

To achieve this, see 4112383d12d589c6d24406210939e55f8e6246a2  @ line 56-63.  A URL value preprocessor is used to extract the current endpoint's arguments, and we store these on the variable.  We later read these when calling MenuEntryMixin.url().  If _endpoint_arguments_constructor is given, these arguments are ignored entirely. A later commit brings argument filtering, meaning generated urls which do not take the arguments won't have a query string appended. This is done by querying the argspec of a method at runtime.  d102c05c15341a2573b3324a573598f3367a17e2

## Compatibility for using Flask-Classy with this library. 
Allows registration via menu.register_flaskview(app, classy_view). Within the Flask View, declare endpoints as menu items using menu.classy_menu_item(path, text, **kwargs), where kwargs is any argument that menu.register would take.  It may be a good design choice to rename these two functions to make them less ambiguous to something such as @menu.register_classy_menu(), since @menu.register_menu is the current non-classy alternative. 

## Adds has_visible_child to check if a child is currently visible for a menu item.
Useful for expanding/contracting menus/sub menus. Simply a recursive call, so use sparingly. (if called with recursive=True)

This pull request includes documentation in the docs folder and has been tested + has additional test cases. Brings coverage up to 99%

jirikuncar commented on docs/index.rst:
```diff
@@ -207,6 +207,45 @@ with 3 items while processing a request for `/social/list`.
 
+
+Flask-Classy
+===
```
please make it the same length like the line above

jirikuncar commented on docs/index.rst:
```diff
@@ -207,6 +207,45 @@ with 3 items while processing a request for `/social/list`.
+
+.. code-block:: python
+
+    from flask.ext.classy import FlaskView
```
for Flask 1.0 we should rather use `flask_<extension>` hence `from flask_classy import ...` and `import flask_menu as menu`

jirikuncar commented on docs/index.rst:
```diff
@@ -207,6 +207,45 @@ with 3 items while processing a request for `/social/list`.
+            pass
+
+
+As you can see, instead of using the `@menu.register_menu` decorator, we use classy_menu_item. All usage is otherwise the same to `register_menu`, however you do not need to provide reference to the blueprint/app.
```
please try to keep the line length reasonable

jirikuncar commented on tests/test_core.py:
```diff
@@ -304,3 +310,380 @@ def test():
+        pass
+
+    def test_classy_endpoint_on_blueprint(self):
+        from flask.ext.classy import FlaskView
```
`s/flask\.ext\.classy/flask_classy/`

nickw444 commented:
@jirikuncar Just pushed the changes you requested. Let me know if there's anything else. Cheers.

jirikuncar commented:
@nickw444 we are also trying to keep our git history well organized. Can you look to our commit message style and squash some of your commits? Thanks

--
PS: I will have closer look to this PR next week. Are you in hurry for integration?

jirikuncar commented on docs/index.rst:
```diff
@@ -207,6 +207,47 @@ with 3 items while processing a request for `/social/list`.
+.. code-block:: python
+
+
+    from flask.ext import menu
```
please make the replacement `s/flask.ext./flask_/` globally ...

nickw444 replied:
@jirikuncar Whilst i'm at it, would you like this done for the entire document, rather than just my additions? (Not sure if you mean globally by my changes, or the entire document)

jirikuncar commented on setup.py:
```diff
@@ -52,7 +52,8 @@ def run_tests(self):
     'pytest>=2.6.1',
-    'coverage<4.0a1'
+    'coverage<4.0a1',
+    'flask-classy>=0.6.10'
```
this dependency should be optional extra

jirikuncar replied:
do you need it also in `extras_require = { ..., 'classy': ['flask-classy>=0.6.10]}`?

jirikuncar replied:
please add comma `,` at the end of line

jirikuncar commented on flask_menu/__init__.py:
```diff
@@ -308,6 +342,110 @@ def _register_menu_item():
+                    )
+
+
+def classy_menu_item(path, text, **kwargs):
```
does it really depends on `classy`? can we find some more generic name?

nickw444 replied:
It would be nice to find a more generic name, however the URL endpoints that classy generates (and this implementation relies on) probably aren't too portable. Classy generates it's endpoints as 'Classname:methodname', see https://github.com/inveniosoftware/flask-menu/pull/33/files#diff-3839e0a4ca06ff71b86a56b51eafdf11R373.

jirikuncar commented:
- [ ] Please add yourself to `AUTHORS`.
- [ ] Please reply if you agree to donate this (and any future) contributions to *Flask-Menu* package under the terms of the Revised BSD Licence.

Thank you!

nickw444 commented:
No rush for integration, i'm just using pip at this point in time to clone my fork. The sooner the better though, keeping upstream synced isn't always fun. 

I've flattened the git history and improved the commit messages (that was difficult). 

Before it gets merged we should come to a decision about the following:
[] Finding a more generic name - does it have to be called classy_menu_item/register_flaskview. The implementation does depend specifically on classy and the way it formats it's endpoints, so we should think about this. In my opinion, it is appropriate, maybe just a better wording?
[] Will I globally rename flask.ext to flask_ for the entire documentation, or just keep it as it is now (where only my new section uses the flask_ convention)

------
Additionally; 

[x] Added myself to `AUTHORS`
[x] Yes, I agree to donating these (and future) contributions to Flask-Menu under the included revised BSD license.

jirikuncar commented on flask_menu/__init__.py:
```diff
@@ -90,6 +97,7 @@ def _active_when(self):
 
     def register(self, endpoint, text, order=0,
                  endpoint_arguments_constructor=None,
+                 expected_args=[],
```
please use `None` as default argument - see http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

jirikuncar commented:
@nickw444 can you split your contribution to general `menu` improvements and new `classy.py` where you can introduce new decorator? You can also move there most of the documentation and reference it in `docs/index.rst` with `.. automodule:: flask_menu.classy`. What do you think?

nickw444 commented:
@jirikuncar Changes have been implemented, Also moved flask-classy to its own section in the API docs.

nickw444 commented:
@jirikuncar  Can you report on the status of this merge request? Thanks

jirikuncar commented on AUTHORS:
```diff
@@ -13,3 +13,4 @@ Contact us at `info@invenio-software.org <mailto:info@invenio-software.org>`_
 * Eirini Psallida <eirini.psallida@cern.ch>
 * Florian Merges <fmerges@fstarter.org>
 * Marco Neumann <marco@crepererum.net>
+* Nick Whyte <nick@nickwhyte.com>
```
please add `\n`

jirikuncar commented on flask_menu/classy.py:
```diff
@@ -0,0 +1,102 @@
+"""Add support for Flask-Classy."""
```
Please include license header.

jirikuncar commented:
Please rebase, remove merge commit, squash related commits and amend commit messages (remove dots from first line, use line separated bullet points for commit message details - see git log - in order to help up autogenerate release notes). Thanks

nickw444 commented:
I have added the required changes. 

Build is failing due to Travis using latest pip, and the travis.yml file using the flag `--upgrade pip`.

tiborsimko commented:
> Build is failing due to Travis using latest pip, and the travis.yml file using the flag --use-mirrors. 

Oops, indeed.  Fix in PR #34.

jirikuncar commented:
@nickw444 you can rebase now. Please make sure that first line in commit message is max. 50 chars long and bullet points have max. 72 chars. Thanks

jirikuncar commented on tests/test_core.py:
```diff
@@ -137,9 +145,9 @@ def always():
         @register_menu(self.app, 'normal', 'Normal')
-        def normal():
+        @self.app.route('/normal/<int:id>/')
+        def normal(id):
```
Can you please add new test instead of modifying the existing one? Thank you!

nickw444 replied:
Sure, i'll get onto that this week

nickw444 commented:
@jirikuncar Apologies for the delay. I have added the new test as opposed to modifying the existing one. I have also changed the commit description to meet your requirements. 

Cheers

jirikuncar commented on flask_menu/__init__.py:
```diff
@@ -90,14 +97,19 @@ def _active_when(self):
 
     def register(self, endpoint, text, order=0,
                  endpoint_arguments_constructor=None,
+                 expected_args=None,
```
just for better backward compatibility, can you move it after `visible_when` argument?

jirikuncar commented on flask_menu/__init__.py:
```diff
@@ -90,14 +97,19 @@ def _active_when(self):
                  visible_when=None,
                  **kwargs):
         """Assign endpoint and display values."""
+        if not expected_args:
```
`expected_args = expected_args or []`

jirikuncar commented on flask_menu/__init__.py:
```diff
@@ -90,14 +97,19 @@ def _active_when(self):
         self._endpoint = endpoint
         self._text = text
         self._order = order
+        self._expected_args = expected_args
```
or even better here: `self._expected_args = expected_args or []`

jirikuncar commented on flask_menu/__init__.py:
```diff
@@ -298,6 +335,7 @@ def _register_menu_item():
                 endpoint,
                 text,
                 order,
+                expected_args=expected,
```
Please move it after `visible_when`.

jirikuncar commented on tests/test_core.py:
```diff
@@ -13,7 +13,15 @@
+)
+from flask_menu.classy import (
+    register_flaskview,
+    classy_menu_item
```
you can add `,` here too

jirikuncar commented:
> I have also changed the commit description to meet your requirements.

@nickw444 Thanks a lot for helping us to keep the commit messages consistent. Can you please finish the sentences with a dot `.`. Thanks again.

nickw444 commented:
Changes made, let me know if there's anything else.

jirikuncar commented on flask_menu/__init__.py:
```diff
@@ -240,6 +259,19 @@ def has_active_child(self, recursive=True):
         return False
 
+    def has_visible_child(self, recursive=True):
+        result = False
```
@nickw444 can you add a short docstring here? Thanks

jirikuncar commented on flask_menu/classy.py:
```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Flask-Menu
+# Copyright (C) 2013, 2014, 2015 CERN.
```
There should be only current year as it's new file.

jirikuncar commented:
@nickw444 I have seen just two little issues in the code and an extra dot on first line in commit message.

--
PS: I take the last commit message problem on me as I was not very clear. We have a document about commit messages in our original repo http://invenio.readthedocs.org/en/latest/developers/git-workflow.html?highlight=git#r2-remarks-on-commit-log-messages and I see that we should extract the most important parts to `CONTRIBUTING.rst` in all ours repos. Thank you for your great contribution! Cheers

nickw444 commented:
@jirikuncar - Just made those last two changes and fixed the commit message

Cheers

jirikuncar commented:
@nickw444 everything is ok now. Can you please rebase on top of `master` (8769194). Thanks

nickw444 commented:
Rebased on top of 8769194. Cheers
